### PR TITLE
feat: Soft lock an app if the certificate wasn't enrolled during the grace period

### PR DESCRIPTION
**Pull request by KaterinaWire**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6804" title="WPB-6804" target="_blank">Sub-taskWPB-6804</a>  [iOS] Soft blocking an app if the certificate is missing within the grace period
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This PR introduces a new app state: `pendingCertificateEnroll`.
In this state the app should be blocked and I reuse the BlockerViewController to implement this.

**When:** If a team admin has enabled E2EI in team settings and the user continued to postpone enrollment. After the grace period finishes, the app should be blocked.

Once the certificate enrollmend is complete, the app will be transferred to an authenticated state.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift [hunk 1]
```diff
@@ -531,6 +534,9 @@ public final class SessionManager: NSObject, SessionManagerType {
         callCenterObserverToken = WireCallCenterV3.addGlobalCallStateObserver(observer: self)
 
         checkJailbreakIfNeeded()
+        Task {
+            await updateOrEnrollCertificateIfNeeded()
+        }
```
Comment (netbe) [inline wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: question: do we have to do this on init? or can it be moved to a function and call it explicitly?

File: wire-ios/Wire-iOS/Sources/AppStateCalculator.swift [hunk 2]
```diff
@@ -27,6 +27,8 @@ enum AppState: Equatable {
     case unauthenticated(error: NSError?)
     case blacklisted(reason: BlacklistReason)
     case jailbroken
+    case pendingCertificateUpdate
+    case pendingCertificateEnroll
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/AppStateCalculator.swift]:
```suggestion
    case certificateUpdateRequired
    case certificateEnrollmentRequired
```

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift [hunk 3]
```diff
@@ -57,6 +57,10 @@ public protocol SessionManagerDelegate: AnyObject, SessionActivationObserver {
     func sessionManagerDidFailToLoadDatabase(error: Error)
     func sessionManagerDidBlacklistCurrentVersion(reason: BlacklistReason)
     func sessionManagerDidBlacklistJailbrokenDevice()
+    func sessionManagerWillUpdateCertificate()
+    func sessionManagerWillEnrollCertificate()
+    func sessionManagerDidUpdateCertificate(for activeSession: UserSession?)
```
Comment (typfel) [inline wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]:
```suggestion
    func sessionManagerRequireCertificateUpdate()
    func sessionManagerRequireCertificateEnrollment()
    func sessionManagerDidUpdateCertificate(for activeSession: UserSession?)
```

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift [hunk 4]
```diff
@@ -1135,6 +1139,18 @@ public final class SessionManager: NSObject, SessionManagerType {
         }
     }
 
+    private func updateOrEnrollCertificateIfNeeded() async {
```
Comment (typfel) [inline wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]:
```suggestion
    private func requestCertificateUpdateOrEnrollIfNeeded() async {
```

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession+Certificate.swift [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+extension ZMUserSession {
+
+    var needsToUpdateCertificate: Bool {
```
Comment (typfel) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession+Certificate.swift]: This isn't very testable, can we move this to a `NeedsToUpdateCertificateUseCase`?

**Review by typfel — CHANGES_REQUESTED** [review]

**Review by typfel — APPROVED** [review]

**Review by David-Henner — APPROVED** [review]

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift (hunk 1 above)
Comment (KaterinaWire) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: this was removed

**Pull request merged**